keg_bouncer/db/mixins.py

```diff
@@ -0,0 +1,76 @@
+import sqlalchemy.orm as saorm
+
+import flask_user
+
+from . import (
+    UserGroup, joined_permission_query, make_link,
+    user_group_permission_map, user_group_bundle_map
+)
+
+
+class UserMixin(flask_user.UserMixin):
+    """A mixin that adds permission facilities to a SQLAlchemy declarative user entity.
+
+    A class which mixes this in must provide the following:
+        * An 'id' field which represents the primary key. It may have any type. If you do not
+          want to call it `id`, you can provide an `id` field as `synonym_for` for your actual
+          primary key.
+    """
+
+    # Instances will override this when caching is needed.
+    _cached_permissions = None
+
+    @declared_attr
+    def id(cls):
+        raise NotImplementedError('You must create an "id" field or a synonym_for it.')
+
+    @declared_attr
+    def user_groups(cls):
+        return saorm.relationship(UserGroup,
+                                  secondary=cls.user_user_group_map,
+                                  cascade='all',
+                                  passive_deletes=True,
+                                  backref='users')
+
+    @declared_attr
+    def user_user_group_map(cls):
+        """A linking (mapping) table between users and user groups."""
+        return make_link('user_user_group_map',
+                         'user_id', cls.id,
+                         'user_group_id', UserGroup.id)
+
+    def get_all_permissions(self):
+        """Get all permissions that are joined to this User, whether directly, through
+           permission bundles, or through user groups."""
+        return frozenset(joined_permission_query().join(
+            self.user_user_group_map,
+            sa.or_(
+                self.user_user_group_map.c.user_group_id
+                    == user_group_permission_map.c.user_group_id,  # noqa
+                self.user_user_group_map.c.user_group_id
+                    == user_group_bundle_map.c.user_group_id  # noqa
+            )
+        ).filter(
+            self.user_user_group_map.c.user_id == self.id,
+        ))
+
+    def get_all_permissions_cached(self):
+        """Same as `get_all_permissions` but caches the result and always returns
+        the same result for this instance of the class.
+
+        This is useful when you want to maximize speed and minimize hits to the database."""
+        if self._cached_permissions is None:
+            self._cached_permissions = self.get_all_permissions()
+        return self._cached_permissions
+
+    def has_permission_token(self, token):
+        """Returns True IFF the given token is present in the user's permission set."""
+        return token in [x.token for x in self.get_all_permissions()]
+
+    def has_permission_token_cached(self, token):
+        """Same as `has_permission_token` except operates on the permission cache."""
+        return token in [x.token for x in self.get_all_permissions()]
```

Comment:
So my approach is mainly governed by the idea of avoiding "future-proofing."

  * I'd rather always cache or never cache. Letting someone "expire" or "refresh" the cache seems like a *potential* future need, but I don't need it now. We may *never* need it, in which case it's unnecessary complexity. Mutation is almost always more complex. (I'm not sure what "handsy" means, but mutating an object that is likely global seems pretty "handsy", if I can guess the meaning). 
  * I intentionally avoided offering methods that check against multiple tokens. The reason is because we both agree that code should not be making "logical deductions" about permissions. In other words, code should never say `if user.has_any('view', 'edit')` because that's making some sort of logical connection between `'view'` and `'edit'`. If they need to be checked in the same place, then we likely need a new permission.